### DISCUSSION TITLE
[bug] [fix] HOW to DISABLE Yeti-RBL COMPLETELY? + Temp Solution for slow roundcube.

### DISCUSSION THREAD

**Discussion by z0lo13** [YetiForceCompany/YetiForceCRM#14656] [General]
Hello Yeti Team,

Thank you for such an amazing project. 

We understand that introducing such a module as RBL is a very nice feature to have, however because of this integration the performance of loading emails and displaying them is dramatically low as pointed out by @mariuszkrzaczkowski in a few issues, as a solution for this is to wait until 6.2.0 which will have an option to turn off the RBL check.

However, today the Roundcube on all yeti instances (10 different instances, with +- different hardware) went very slow, we found out that it was due to attachments, specifically the ones more than 10 MB, some of the instances just hang up. The roundcube works blazingly fast after disabling yetiforce plugin for roundcube. At the same time from issues i have found out the possible solution to comment out the line 32 in yetiforce.php file. It does helps, the performance increase in about 5 times, however, it hangs the instances if an email with heavy attachments is loaded. 

I have commented out mail analysis and all similar functions which send requests to RBL,  however, it is definitely not the best way to do it.
Please expand the functionality of disabling the RBL service, in order for us to  be able to disable it completely (0 Request sent to RBL server)

After debugging below are the results:

**TEMP SOLUTION: or anyone having same issue**
1. The :32 line is commented out in yetiforce.php
2. After debugging I have determined that the issue of why the instance hangs when it contains a large attachment. 
    modules/OSSMail/views/MailActionBar.php :35 CheckMailExist function

3. as a result refers modules/OSSMailView/models/Record.php :321

4. Next investigation pointed out that OSSMail_Record_Model::getMail function does download attachments, and FullMode is enabled by default.  See below code as a reference. 
modules/OSSMail/models/Record.php
```
/**
	 * @param resource $mbox
	 * @param int      $id
	 * @param int      $msgno
	 * @param bool     $fullMode
	 *
	 * @return bool|\OSSMail_Mail_Model
	 */
	public static function getMail($mbox, $id, $msgno = false, bool $fullMode = true)
	{
		if (!$msgno) {
			\App\Log::beginProfile(__METHOD__ . '|imap_msgno', 'Mail|IMAP');
			$msgno = imap_msgno($mbox, $id);
			\App\Log::endProfile(__METHOD__ . '|imap_msgno', 'Mail|IMAP');
		}
		if (!$id) {
			\App\Log::beginProfile(__METHOD__ . '|imap_uid', 'Mail|IMAP');
			$id = imap_uid($mbox, $msgno);
			\App\Log::endProfile(__METHOD__ . '|imap_uid', 'Mail|IMAP');
		}
		if (!$msgno) {
			return false;
		}
		\App\Log::beginProfile(__METHOD__ . '|imap_headerinfo', 'Mail|IMAP');
		$header = imap_headerinfo($mbox, $msgno);
		\App\Log::endProfile(__METHOD__ . '|imap_headerinfo', 'Mail|IMAP');
		$messageId = '';
		if (property_exists($header, 'message_id')) {
			$messageId = $header->message_id;
		}
		$mail = new OSSMail_Mail_Model();
		$mail->set('header', $header);
		$mail->set('id', $id);
		$mail->set('Msgno', $header->Msgno);
		$mail->set('message_id', $messageId ? \App\Purifier::purifyByType($messageId, 'MailId') : '');
		$mail->set('to_email', \App\Purifier::purify($mail->getEmail('to')));
		$mail->set('from_email', \App\Purifier::purify($mail->getEmail('from')));
		$mail->set('reply_toaddress', \App\Purifier::purify($mail->getEmail('reply_to')));
		$mail->set('cc_email', \App\Purifier::purify($mail->getEmail('cc')));
		$mail->set('bcc_email', \App\Purifier::purify($mail->getEmail('bcc')));
		$mail->set('subject', isset($header->subject) ? \App\TextParser::textTruncate(\App\Purifier::purify(self::decodeText($header->subject)), 65535, false) : '');
		$mail->set('date', date('Y-m-d H:i:s', $header->udate));
		if ($fullMode) {
			$structure = self::getBodyAttach($mbox, $id, $msgno);
			$mail->set('body', $structure['body']);
			$mail->set('attachments', $structure['attachment']);

			$clean = '';
			\App\Log::beginProfile(__METHOD__ . '|imap_fetch_overview', 'Mail|IMAP');
			$msgs = imap_fetch_overview($mbox, $msgno);
			\App\Log::endProfile(__METHOD__ . '|imap_fetch_overview', 'Mail|IMAP');

			foreach ($msgs as $msg) {
				\App\Log::beginProfile(__METHOD__ . '|imap_fetchheader', 'Mail|IMAP');
				$clean .= imap_fetchheader($mbox, $msg->msgno);
				\App\Log::endProfile(__METHOD__ . '|imap_fetchheader', 'Mail|IMAP');
			}
			$mail->set('clean', $clean);
		}
		return $mail;
	}
```

5. I don't know is it a typo or intend use but line 323 of "modules/OSSMailView/models/Record.php" does not turn off the full mode and as a result, the attachments being downloaded. 
Replacing
  
`$mail = OSSMail_Record_Model::getMail($mbox, $uid, false);`   
with 
`$mail = OSSMail_Record_Model::getMail($mbox, $uid, false, false);`

6. Fixed the problem when the system hangs when previewing emails with large attachments, however, i am not sure if the other functions will be impacted by disabling "Full Mode"

7. What are the possible outcomes if fullMode disabled while calling checkMailExist from OSSMailView_Record_Model? Will the other functions won't work correctly? As I was not able to find any use of the attachment portion for checking if mail exists in the database.

I hope the above debugging and info will help you improve the yetiforce plugin for roundcube as well as RBL checks.

**Comment by waran70** [comment]
Don't disable! I think RBL pay commission ;-)
https://github.com/YetiForceCompany/YetiForceCRM/issues/14181

**Comment by z0lo13** [comment]
> Don't disable! I think RBL pay commission ;-)
> #14181

Your comment does not make sense, as 6.2.0 and dev version will have an option to disable the RBL check.
https://github.com/YetiForceCompany/YetiForceCRM/issues/14489#issuecomment-813052569

Also, I don't really know what was the point of the conversation in #114181, you've sent them an email, your IP got blocked for spam, @bpabiszczak offered to unblock your IP free of charge, (NO you don't need to pay 5Euro in cases when it was not intentional), which you denied and just said that Toptal freelance support is better :-D

I guess you just bored man.

P.S.
It is a point to have it enabled, but not when the email client non-usable if the attachment is more than 15 MB.

**Comment by waran70** [comment]
@z0lo13: It was response to email from yetiforce! Nobody spammed 'em, or even sent cold email. Totally legit and clean of any spam records address was dumped in RBL and we have been blackmailed with 5 Euro to get IP back on track, so yetiforce stuff can in their kindness read our business inquiry. 
In the next step offering me unblocking my IP for free is a sick thing. 
We talking about business case where customers (like mine)  don't bother with guys who creating blockers for them and who extending time of resolving case. 

Are you sure I am bored man? Little analysis and you got right picture.

**Reply by z0lo13** [reply]
@waran70 
1. Believe me, other customers, **Unlike you** will find a way to contact the correct person/ department if it's urgent and they won't be discussing that on Github, yetiforce does have an office and phone numbers.
2. As with any anti Spam Software its does misfires based on multiple factors (Headers, SPF, dkim, and so on and on.) If you think you are sending totally legit emails it does not mean the Anti Spam software will think the same way, it is not human but just the script which checks for specifics, and obviously one of that specific checks your email didn't pass.
3. Of course I am not sure that you are bored; maybe you have some adhd issues... I have no idea... and I don't really care, how you will call it.
4. If you have nothing to comment on the issue, which will help to solve it, don't waste anyone's time.

**Reply by JohnMaverick** [reply]
Whiles he tried to make little fun out of RBL you are going to be very serious boy. Enjoy than.

**Reply by noelhunter** [reply]
Since the upgrade to 6.3.0, our server has been freezing up whenever a client opens an email that triggers an RBL check.  We disabled the RBL settings:

Check the sender on the email list in the mail client  
Check the sender in the message preview in the mail client 
Accept automatically
Send automatically

but the check is STILL being performed.  It appears to execute so slowly, and possibly multiple times, so that it exhausts the pool of Nginx workers.  If we edit the OSS mail file /config/Modules/OSSMail.php to disable the yetiforce plugin for Roundcube, the problem is resolved (but none of the yetiforce enhancements are available).

Is there a way to stop these checks against RBL from hanging the server, and still keep the yetiforce Roundcube module enabled?

**Reply by mariuszkrzaczkowski** [reply]
try this
https://doc.yetiforce.com/code/classes/Config-Components-Mail.html#property_rcListAcceptAutomatically
config/Components/Mail.php
```php
	/** Check the sender on the email list in the mail client */
	public static $rcListCheckRbl = false;

	/** Check the sender in the message preview in the mail client */
	public static $rcDetailCheckRbl = false;
```

**Reply by noelhunter** [reply]
Thanks, we have set those options in config/Components/Mail.php, but the YF-SOC icons still appear on some incoming messages, and the check still appears to perform. (this is on the 6.3.0 stable dist)

**Reply by mariuszkrzaczkowski** [reply]
give me more details,
give the version number of the mail client, the file config/version.php  `'lib_roundcube' => '???',`

**Reply by noelhunter** [reply]
return [
        'appVersion' => '6.3.0',
        'patchVersion' => '2021.12.10',
        'lib_roundcube' => '0.2.3',
];

**Reply by mariuszkrzaczkowski** [reply]
change the version lib_roundcube to 0.2.10, update the library, and see if it works better

**Reply by noelhunter** [reply]
Thanks, I made the update, and will reply if the problem is solved (or made worse)

**Reply by noelhunter** [reply]
That seemed to help the slowdown.  However, the YF-SOC icons are still appearing in messages with the settings manually disabled in both the UI and config file.

**Reply by mariuszkrzaczkowski** [reply]
I uploaded the patch, set the 0.2.11 version

**Reply by noelhunter** [reply]
Thank you.  0.2.11 fixed the problem, and the YF-SOC icons / slowness are no linger appearing when the RBL settings are set to off.

**Reply by simple-AMS** [reply]
@mariuszkrzaczkowski 

How do you disable the RBL Mailservers and YF-SOC icons inside of Roundcube altogether?

**Reply by mariuszkrzaczkowski** [reply]
https://github.com/YetiForceCompany/YetiForceCRM/discussions/14656#discussioncomment-2069457  will try it
